crates/bevy_ecs/src/reflect/map_entities.rs

```diff
@@ -12,53 +14,73 @@ use bevy_reflect::FromType;
     ///
     /// Be mindful in its usage: Works best in situations where the entities in the [`EntityHashMap<Entity>`] are newly
     /// created, before systems have a chance to add new components. If some of the entities referred to
-    /// by the [`EntityHashMap<Entity>`] might already contain valid entity references, you should use [`map_entities`](Self::map_entities).
+    /// by the [`EntityHashMap<Entity>`] might already contain valid entity references, you should use [`map_world_entities`](Self::map_world_entities).
     ///
     /// An example of this: A scene can be loaded with `Parent` components, but then a `Parent` component can be added
-    /// to these entities after they have been loaded. If you reload the scene using [`map_all_entities`](Self::map_all_entities), those `Parent`
+    /// to these entities after they have been loaded. If you reload the scene using [`map_all_world_entities`](Self::map_all_world_entities), those `Parent`
     /// components with already valid entity references could be updated to point at something else entirely.
-    pub fn map_all_entities(&self, world: &mut World, entity_map: &mut EntityHashMap<Entity>) {
-        SceneEntityMapper::world_scope(entity_map, world, self.map_all_entities);
+    #[deprecated = "map_all_world_entities doesn't play well with Observers. Use map_entities instead."]
+    pub fn map_all_world_entities(
+        &self,
+        world: &mut World,
+        entity_map: &mut EntityHashMap<Entity>,
+    ) {
+        SceneEntityMapper::world_scope(entity_map, world, self.map_all_world_entities);
     }
 
     /// A general method for applying [`MapEntities`] behavior to elements in an [`EntityHashMap<Entity>`]. Unlike
-    /// [`map_all_entities`](Self::map_all_entities), this is applied to specific entities, not all values
+    /// [`map_all_world_entities`](Self::map_all_world_entities), this is applied to specific entities, not all values
     /// in the [`EntityHashMap<Entity>`].
     ///
     /// This is useful mostly for when you need to be careful not to update components that already contain valid entity
-    /// values. See [`map_all_entities`](Self::map_all_entities) for more details.
-    pub fn map_entities(
+    /// values. See [`map_all_world_entities`](Self::map_all_world_entities) for more details.
+    #[deprecated = "map_world_entities doesn't play well with Observers. Use map_entities instead."]
+    pub fn map_world_entities(
         &self,
         world: &mut World,
         entity_map: &mut EntityHashMap<Entity>,
         entities: &[Entity],
     ) {
         SceneEntityMapper::world_scope(entity_map, world, |world, mapper| {
-            (self.map_entities)(world, mapper, entities);
+            (self.map_world_entities)(world, mapper, entities);
         });
     }
 }
 
-impl<C: Component + MapEntities> FromType<C> for ReflectMapEntities {
+impl<C: Component + MapEntities + Reflect + FromReflect> FromType<C> for ReflectMapEntities {
```

Comment:
This is slightly more restrictive than previously, but I'm not sure that it really matters since anything reflecting `MapEntities` should already be `Reflect + FromReflect`.

Reply:
I think it's fine. Ideally we'd only use `Reflect` and do a similar trick as `ReflectComponent`, but I’m not sure that's possible without world access.

Reply:
World access shouldn't actually be a problem, at least for the two places we're currently using this. I actually had a `&mut World` parameter previously before swapping the `EntityHashMap<Entity>` for the `&mut dyn DynEntityMapper`.

Reply:
Yeah, [works just fine](https://github.com/bevyengine/bevy/commit/7ddb1c2078f3c2ec103b024366e8b996dc37b169). I could go either way :shrug: